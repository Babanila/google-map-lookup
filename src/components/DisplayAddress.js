import React from 'react'
import { useHistory } from 'react-router-dom'
import { cx, css } from 'emotion'
import { ErrorPage, LoadingPage, SingleButton } from './ReusableComponents'
import { simpleFetcher } from '../utils/helpers'
import MapComponent from './MapComponent'

const displayAddressDiv = css`
  margin-top: 6em;
  width: 100%;
  height: 600px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  z-index: 10;
`

const backButtonDiv = css`
  width: 100%;
  max-width: 500px;
  height: 3em;
  display: flex;
  align-items: center;
  justify-content: flex-end;
  background: transparent;
`

const backBtnStyle = css`
  padding: 10px;
  font-size: 1em;
  color: #ffffff;
  background-color: #36a7fe;
  border: 2px solid #36a7fe;
  border-radius: 5px;
  outline: none;
  &:hover {
    cursor: pointer;
    color: #36a7fe;
    background-color: #ffffff;
  }
`

const cache = {}

export function DisplayAddress({ match }) {
  const [locationData, setLocationData] = React.useState({})
  const [loading, setLoading] = React.useState(true)
  const [error, setError] = React.useState('')

  const { postcode } = match.params
  const addressUrl = `https://maps.googleapis.com/maps/api/geocode/json?address=${postcode}&key=${process.env.API_KEY}`

  React.useEffect(() => {
    getAddressDetails(addressUrl)
  }, [postcode, addressUrl])

  const getAddressDetails = async (url) => {
    try {
      if (cache[url]) {
        const data = await cache[url]
        setLocationData(data)
        setLoading(false)
      } else {
        const { data } = await simpleFetcher(url)
        cache[url] = data
        setLocationData(data)
        setLoading(false)
      }
    } catch (error) {
      setError(error)
      setLoading(false)
    }
  }

  const history = useHistory()
  const handleReturnBackHome = () => history.push(`/`)

  if (error) return <ErrorPage message={error} />
  if (loading) return <LoadingPage />

  return (
    <div className={cx(displayAddressDiv)} onClick={() => {}}>
      <div className={cx(backButtonDiv)}>
        <SingleButton btnName="Back" btnClick={handleReturnBackHome} btnStyles={backBtnStyle} />
      </div>
      <MapComponent addressData={locationData} />
    </div>
  )
}

export default DisplayAddress
