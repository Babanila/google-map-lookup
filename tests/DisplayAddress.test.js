import React from 'react'
import renderer from 'react-test-renderer'
import { mount } from 'enzyme'
import { MemoryRouter } from 'react-router-dom'
import DisplayAddress from '../src/components/DisplayAddress'

describe('Check DisplayAddress Component', () => {
  let sampleData, match

  beforeEach(() => {
    match = { params: { postcode: 90453 } }
    sampleData = {
      results: [
        {
          formatted_address: '90453 NÃ¼rnberg, Deutschland',
          geometry: {
            bounds: {
              northeast: { lat: 49.4012891, lng: 11.06465 },
              southwest: { lat: 49.35221, lng: 10.9925109 }
            },
            location: {
              lat: 49.3757248,
              lng: 11.0346054
            },
            location_type: 'APPROXIMATE',
            viewport: {
              northeast: { lat: 49.4012891, lng: 11.06465 },
              southwest: { lat: 49.35221, lng: 10.9925109 }
            }
          },
          place_id: 'ChIJ26TczgNRn0cRwEbF9DXaHhw',
          types: ['postal_code']
        }
      ],
      status: 'OK'
    }
  })

  it('should renders component without crashing', () => {
    const tree = renderer.create(<DisplayAddress match={match} />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it('should render DisplayAddress component without data', () => {
    const DisplayAddressWrapper = mount(
      <MemoryRouter initialentries="{['/']}">
        <DisplayAddress match={match} />
      </MemoryRouter>
    )

    const actualDiv = DisplayAddressWrapper.find('MemoryRouter')
    const actualDiv1 = actualDiv.find('Router')
    const actualComponent = actualDiv1.find('DisplayAddress')
    const actualDiv3 = actualComponent.find('LoadingPage')
    expect(DisplayAddress).toBeDefined()
    expect(actualDiv.length).toEqual(1)
    expect(actualDiv1.length).toEqual(1)
    expect(actualComponent.length).toEqual(1)
    expect(actualDiv3.length).toEqual(1)
  })

  it('should render DisplayAddress component data', () => {
    const setLocationData = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setLocationData({ ...sampleData }))

    const setLoading = React.useState
    jest.spyOn(React, 'useState').mockImplementationOnce(() => setLoading(false))

    const DisplayAddressWrap = mount(
      <MemoryRouter initialentries="{['/']}">
        <DisplayAddress match={match} />
      </MemoryRouter>
    )

    const actual1 = DisplayAddressWrap.find('div')
    const buttonComponent = DisplayAddressWrap.find('SingleButton')
    const mapComponent = DisplayAddressWrap.find('Memo(MapComponent)')
    expect(actual1.length).toEqual(4)
    expect(buttonComponent.length).toEqual(1)
    expect(mapComponent.length).toEqual(1)
  })
})
