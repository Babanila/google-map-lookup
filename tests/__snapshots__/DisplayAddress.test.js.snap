// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Check DisplayAddress Component should renders component without crashing 1`] = `
<h3
  className="css-a4inhu"
>
  Loading ...
</h3>
`;
